[FLINK-18735][table] Add support for more types to DataGen source

sjwiesman opened the pull request:
## What is the purpose of the change

Updates the DataGen source to support most types. This enables using DataGen in conjunction with Flink's LIKE clause as it currently does not support overwriting physical columns with computed columns. 

## Brief change log

I refactored the table factory and added support for all types except RAW and STRUCTURED. See commit messages for a full list of changes. I also added support for creating bounded tables which is useful when prototyping a query. 


## Verifying this change

New unit tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

JingsongLi commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/DataGeneratorSource.java:
```diff
@@ -77,8 +102,11 @@ public void run(SourceContext<T> ctx) throws Exception {
+				if (isRunning && generator.hasNext() && (numberOfRows == null || outputSoFar < toOutput)) {
 					synchronized (ctx.getCheckpointLock()) {
+						if (numberOfRows != null) {
+							outputSoFar++;
```
I have an idea: print `outputSoFar` in `close`. It is good for debugging.

JingsongLi commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/DataGeneratorSource.java:
```diff
@@ -77,8 +102,11 @@ public void run(SourceContext<T> ctx) throws Exception {
-				if (isRunning && generator.hasNext()) {
+				if (isRunning && generator.hasNext() && (numberOfRows == null || outputSoFar < toOutput)) {
 					synchronized (ctx.getCheckpointLock()) {
+						if (numberOfRows != null) {
```
Remove this if?

JingsongLi commented on flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/RandomGeneratorVisitor.java:
```diff
@@ -0,0 +1,362 @@
+		Set<ConfigOption<?>> options = keyContainer.getOptions();
+		options.addAll(valContainer.getOptions());
+
+		return DataGeneratorContainer.of(
```
Take a look to the comments of `RowData`. The data strutures should be:
```
 * +--------------------------------+-----------------------------------------+
 * | Row                            | {@link RowData}                         |
 * +--------------------------------+-----------------------------------------+
 * | ARRAY                          | {@link ArrayData}                       |
 * +--------------------------------+-----------------------------------------+
 * | MAP / MULTISET                 | {@link MapData}                         |
 * +--------------------------------+-----------------------------------------+
```

JingsongLi commented on docs/dev/table/connectors/datagen.md:
```diff
@@ -29,25 +29,24 @@ under the License.
 
-- Random generator is the default generator, you can specify random max and min values. For char/varchar/string, the length can be specified. It is a unbounded generator.
-- Sequence generator, you can specify sequence start and end values. It is a bounded generator, when the sequence number reaches the end value, the reading ends.
+Time types are always the local machines current system time.
```
Maybe we can have a table to show all types.
Display the generation strategies they support, and the required parameters?

sjwiesman commented:
@JingsongLi thanks for taking a look. I've updated my PR accordingly.

sjwiesman commented:
@JingsongLi what do you think of the status of this PR?

JingsongLi commented on flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/DataGenVisitorBase.java:
```diff
@@ -0,0 +1,110 @@
+		throw new ValidationException("Unsupported type: " + logicalType);
+	}
+
+	private interface SerializableSupplier<T> extends Supplier<T>, Serializable { }
```
I think lambda is already `Serializable`.

sjwiesman replied:
Its not, the source gets caught in the closure cleaner without this.

JingsongLi replied:
I meant we could do something like `return (Supplier<FileSystem> & Serializable) () -> {...}`, but it is OK for now.

JingsongLi commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/DataGeneratorSource.java:
```diff
@@ -45,18 +62,34 @@
-	public DataGeneratorSource(DataGenerator<T> generator, long rowsPerSecond) {
+	public DataGeneratorSource(DataGenerator<T> generator, String name, long rowsPerSecond, Long numberOfRows) {
 		this.generator = generator;
+		this.name = name;
```
Do we need to add name?

sjwiesman replied:
that's a remnant from an old commit, will remove.

JingsongLi commented on flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DataGeneratorMapper.java:
```diff
@@ -0,0 +1,60 @@
+		return mapper.apply(generator.next());
+	}
+
+	public interface SerializableFunction<A, B> extends Function<A, B>, Serializable {}
```
Ditto.

sjwiesman replied:
same, we need this.

JingsongLi commented on flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/RowDataGenerator.java:
```diff
@@ -0,0 +1,80 @@
+
+	private static final long serialVersionUID = 1L;
+
+	private final DataGenerator[] fieldGenerators;
```
Can we add `?` to `DataGenerator` like `DataGenerator<?>[]`? At least, The compiler will have no warning.

JingsongLi commented on flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java:
```diff
@@ -0,0 +1,64 @@
+
+import org.junit.Test;
+
+public class DataGeneratorConnectorITCase extends StreamingTestBase {
```
You can use `BatchTestBase`.

JingsongLi commented on flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/DataGenTableSource.java:
```diff
@@ -0,0 +1,89 @@
+
+	@Override
+	public ScanRuntimeProvider getScanRuntimeProvider(ScanContext context) {
+		boolean isBounded = numberOfRows == null;
```
Looks like the opposite?

sjwiesman replied:
nice catch

JingsongLi commented on flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java:
```diff
@@ -0,0 +1,64 @@
+	public void testTypes() {
+		tEnv().executeSql(TABLE);
+		tEnv().executeSql(SINK);
+		tEnv().from("datagen_t").executeInsert("sink");
```
You should use: `Lists.newArrayList(tEnv().executeSql("select * from datagen_t").collect())`.
In this way, the types can be verified.

JingsongLi commented on flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/SequenceGeneratorVisitor.java:
```diff
@@ -0,0 +1,177 @@
+	@Override
+	public DataGeneratorContainer visit(DecimalType decimalType) {
+		return DataGeneratorContainer.of(
+			SequenceGenerator.bigDecimalGenerator(
```
Looks like the structures of Decimal, Times are wrong, you can take a look to the comments of `RowData`.

JingsongLi reviewed (changes requested):
I am very sorry for the late response. Left some comments. Feel free to ping me...

sjwiesman commented:
@JingsongLi thank you for the thorough. Review. I've corrected the types and updated the test to validate them. If you don't have any other comments I'll merge on green.

JingsongLi commented on flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/RandomGeneratorVisitor.java:
```diff
@@ -0,0 +1,337 @@
+		};
+	}
+
+	private static class BigDecimalRandomGenerator implements DataGenerator<DecimalData> {
```
I think it can be named `DecimalDataRandomGenerator`
BigDecimal makes confused

JingsongLi reviewed (approved):
Thanks for update, Looks good to me except time type generator, and left some minor comments.

sjwiesman commented:
Sorry for the long back and forth, just pushed a fix addressing your comments.
